llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,359 @@
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "llvm/Support/raw_ostream.h"
+
+#include <queue>
+#include <unordered_map>
+#include <unordered_set>
+
+using namespace llvm;
+
+namespace {
+
+class SYCLPropagateAspectUsageLegacyPass : public ModulePass {
+public:
+  static char ID; // Pass identification, replacement for typeid
+  SYCLPropagateAspectUsageLegacyPass() : ModulePass(ID) {
+    initializeSYCLPropagateAspectUsageLegacyPassPass(
+        *PassRegistry::getPassRegistry());
+  }
+
+  bool runOnModule(Module &M) override {
+    ModuleAnalysisManager MAM;
+    auto PA = Impl.run(M, MAM);
+    return !PA.areAllPreserved();
+  }
+
+private:
+  PropagateAspectUsagePass Impl;
+};
+
+} // namespace
+
+char SYCLPropagateAspectUsageLegacyPass::ID = 0;
+INITIALIZE_PASS(SYCLPropagateAspectUsageLegacyPass, "PropagateAspectUsage",
+                "Propagates aspect usage", false, false)
+
+ModulePass *llvm::createPropagateAspectUsagePass() {
+  return new SYCLPropagateAspectUsageLegacyPass();
+}
+
+namespace {
+
+using AspectsTy = SmallSet<int, 4>;
+using TypesWithAspectsTy = std::unordered_map<Type *, AspectsTy>;
+
+TypesWithAspectsTy GetTypesWithAspectsFromMetadata(Module &M) {
+  NamedMDNode *Node = M.getNamedMetadata("intel_types_that_use_aspects");
+  TypesWithAspectsTy Result;
+  if (!Node)
+    return Result;
+
+  LLVMContext &C = M.getContext();
+  for (size_t i = 0; i != Node->getNumOperands(); ++i) {
+    MDNode *N = Node->getOperand(i);
+    assert(N->getNumOperands() > 1 && "intel_types_that_use_aspect metadata "
+                                      "shouldn't contain empty metadata nodes");
+
+    MDString *TypeName = cast<MDString>(N->getOperand(0));
+    Type *T = StructType::getTypeByName(C, TypeName->getString());
+    if (!T) {
+      // TODO: warn?
+      continue;
+    }
+
+    AspectsTy &Aspects = Result[T];
+    for (size_t j = 1; j != N->getNumOperands(); ++j) {
+      Constant *C = cast<ConstantAsMetadata>(N->getOperand(j))->getValue();
+      Aspects.insert(cast<ConstantInt>(C)->getSExtValue());
+    }
+  }
+
+  return Result;
+}
+
+using TypesEdgesTy = std::unordered_map<Type *, std::vector<Type *>>;
+
+void PropagateAspectsThroughTypes(TypesEdgesTy &Edges, Type *Start,
+                                  AspectsTy &Aspects,
+                                  TypesWithAspectsTy &ResultAspects) {
+  std::unordered_set<Type *> Visited;
+  std::queue<Type *> queue;
+  queue.push(Start);
+  while (!queue.empty()) {
+    Type *T = queue.front();
+    queue.pop();
+    if (Visited.count(T))
+      continue;
+
+    Visited.insert(T);
+    ResultAspects[T].insert(Aspects.begin(), Aspects.end());
+    for (Type *TT : Edges[T]) {
+      if (Visited.count(TT))
+        continue;
+
+      queue.push(TT);
+    }
+  }
+}
+
+TypesWithAspectsTy
+GetTypesWithAspectsFromModule(Module &M, TypesWithAspectsTy &TypesWithAspects) {
```

Comment:
There are different semantics. Argument map consist of known types from metadata information. While returned Map consist of types which we figured out during the search in this function.
The good point in current version is that it is stupid and straightforward because everything is done step by step in main run() function at the end of the file.